core/solver/tfqmr.cpp

```diff
@@ -0,0 +1,248 @@
+    auto theta = Vector::create_with_config_of(alpha.get());
+    auto eta = Vector::create_with_config_of(alpha.get());
+    auto rov = Vector::create_with_config_of(alpha.get());
+
+    auto master_tau =
+        Vector::create(exec->get_master(), 1, dense_b->get_num_cols());
+    auto starting_tau = Vector::create_with_config_of(master_tau.get());
+
+    // TODO: replace this with automatic merged kernel generator
+    exec->run(TemplatedOperation<ValueType>::make_initialize_operation(
+        dense_b, r.get(), r0.get(), u_m.get(), u_mp1.get(), pu_m.get(),
+        Au.get(), Ad.get(), w.get(), v.get(), d.get(), taut.get(),
+        rho_old.get(), rho.get(), alpha.get(), beta.get(), taut.get(),
+        sigma.get(), rov.get(), eta.get(), nomw.get(), theta.get()));
+    // r = dense_b
+    // r0 = u_m = w = r
+    // Ad = d = 0
+    // theta = eta = rov = alpha = beta = sigma = 1.0
+
+    system_matrix_->apply(neg_one_op.get(), dense_x, one_op.get(), r.get());
+    r0->copy_from(r.get());
+    r->compute_dot(r.get(), tau.get());
+    w->copy_from(r.get());
+    u_m->copy_from(r.get());
+    starting_tau->copy_from(tau.get());
+    rho->copy_from(tau.get());
+    rho_old->copy_from(tau.get());
+    taut->copy_from(tau.get());
+    preconditioner_->apply(u_m.get(), pu_m.get());
+    system_matrix_->apply(pu_m.get(), v.get());
+    Au->copy_from(v.get());
+    Ad->copy_from(d.get());
+
+    for (int iter = 0; iter < max_iters_; ++iter) {
+        if (iter % 2 == 0) {
+            r0->compute_dot(v.get(), rov.get());
+            exec->run(TemplatedOperation<ValueType>::make_step_1_operation(
+                alpha.get(), rov.get(), rho.get(), v.get(), u_m.get(),
+                u_mp1.get()));
+            // alpha = rho / rov
+            // u_mp1 = u_m - alpha * v
+        }
+        exec->run(TemplatedOperation<ValueType>::make_step_2_operation(
+            theta.get(), alpha.get(), eta.get(), sigma.get(), Au.get(),
+            pu_m.get(), w.get(), d.get(), Ad.get()));
+        // sigma = (theta^2 / alpha) * eta;
+        // w = w - alpha * Au
+        // d = pu_m + sigma * d
+        // Ad = Au + sigma * Ad
+        w->compute_dot(w.get(), nomw.get());
+        exec->run(TemplatedOperation<ValueType>::make_step_3_operation(
+            theta.get(), nomw.get(), taut.get(), eta.get(), alpha.get()));
+        // theta = nomw / taut
+        // c_mp1 = 1 / (1 + theta)
+        // taut = taut * sqrt(theta) * c_mp1;
+        // eta = c_mp1 * c_mp1 * alpha;
+        exec->run(TemplatedOperation<ValueType>::make_step_4_operation(
+            eta.get(), d.get(), Ad.get(), dense_x, r.get()));
+        // x = x + eta * d
+        // r = r - eta * Ad
+        r->compute_dot(r.get(), tau.get());
+        master_tau->copy_from(tau.get());
+        if (has_converged(master_tau.get(), starting_tau.get(),
+                          rel_residual_goal_)) {
+            break;
+        }
+        if (iter % 2 != 0) {
```

Comment:
Maybe this `if` clause can go into an helper function after combining the two steps 1 and 5 so that you just have to pass in your different inputs to the helper function.